// warp-contracts-pubsub/lib/cjs/appsync/appSync.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.subscribe = exports.publish = exports.initPubSub = void 0;
const api_1 = require("@aws-amplify/api");
const core_1 = require("@aws-amplify/core");
const subscribeDoc = /* GraphQL */ `
  subscription Subscribe($name: String!) {
    subscribe(name: $name) {
      data
      name
    }
  }
`;
const publishDoc = /* GraphQL */ `
  mutation Publish($data: AWSJSON!, $name: String!) {
    publish(data: $data, name: $name) {
      data
      name
    }
  }
`;
function initPubSub() {
    core_1.Amplify.configure({
        aws_appsync_graphqlEndpoint: 'https://6z7ifnfgrfhnxm57ky46bus2su.appsync-api.eu-north-1.amazonaws.com/graphql',
        aws_appsync_region: 'eu-north-1',
        aws_appsync_authenticationType: 'AWS_LAMBDA'
    });
}
exports.initPubSub = initPubSub;
async function publish(name, data, authToken) {
    return api_1.API.graphql((0, api_1.graphqlOperation)(publishDoc, { name, data }, authToken));
}
exports.publish = publish;
async function subscribe(name, next, error) {
    return api_1.API.graphql((0, api_1.graphqlOperation)(subscribeDoc, { name }, 'any')).subscribe({
        next: ({ provider, value }) => {
            next(value.data.subscribe, provider, value);
        },
        error: error || console.dir
    });
}
exports.subscribe = subscribe;
