// @execution-machine/sdk/dist/common/utils/commons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.figureOutContractType = exports.guidGenerator = exports.postRequest = void 0;
const model_1 = require("../model");
const postRequest = (url, body) => {
    return fetch(url, {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(body)
    });
};
exports.postRequest = postRequest;
const guidGenerator = () => {
    const S4 = function () {
        return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    };
    return (S4() + S4() + "-" + S4() + "-" + S4() + "-" + S4() + "-" + S4() + S4() + S4());
};
exports.guidGenerator = guidGenerator;
const figureOutContractType = (input, byType) => {
    let contentType = model_1.ContractType.JS;
    if (input) {
        const lowerCaseInput = input.toLowerCase();
        if (byType) {
            switch (lowerCaseInput) {
                case "wasm":
                    contentType = model_1.ContractType.WASM;
                    break;
                case "evm":
                    contentType = model_1.ContractType.EVM;
                    break;
                case "js":
                    contentType = model_1.ContractType.JS;
                    break;
            }
        }
        else {
            if (lowerCaseInput.endsWith(".js")) {
                contentType = model_1.ContractType.JS;
            }
            else if (lowerCaseInput.endsWith(".wasm")) {
                contentType = model_1.ContractType.WASM;
            }
            else {
                contentType = model_1.ContractType.JS;
            }
        }
    }
    return contentType;
};
exports.figureOutContractType = figureOutContractType;
