// @execution-machine/sdk/dist/test/test-function.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestFunction = exports.createWrite = exports.FunctionType = void 0;
const three_em_0_3_22_1 = require("three-em-0-3-22");
const commons_1 = require("../common/utils/commons");
exports.FunctionType = Object.assign({}, three_em_0_3_22_1.SimulateContractType);
const createWrite = (input, tags, opts, inputIsJson) => {
    return {
        id: opts?.id || (0, commons_1.guidGenerator)(),
        owner: opts?.owner || "@test",
        quantity: opts?.quantity || "0",
        reward: opts?.reward || "0",
        target: opts?.target,
        tags: tags || [],
        block: opts?.block,
        input: inputIsJson ? input : JSON.stringify(input)
    };
};
exports.createWrite = createWrite;
const TestFunction = async (opts) => {
    return await (0, three_em_0_3_22_1.simulateContract)({
        contractId: "",
        maybeContractSource: {
            contractSrc: opts.functionSource,
            contractType: opts.functionType
        },
        contractInitState: JSON.stringify(opts.functionInitState),
        maybeConfig: opts.gatewayConfig,
        maybeSettings: {
            ...(opts.settings || {}),
            'TX_DATE': `${new Date().getTime()}`
        },
        maybeExmContext: opts.exmContext,
        interactions: opts.writes,
        maybeCache: false
    });
};
exports.TestFunction = TestFunction;
